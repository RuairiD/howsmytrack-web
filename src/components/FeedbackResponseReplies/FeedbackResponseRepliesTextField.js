
import React, { useCallback, useEffect, useState, useRef } from "react";
import { useQuery, useMutation } from "react-query";
import axios from "axios";

import { Alert, Button, Checkbox, Col, Input, Row, Spin, Typography } from "antd";
import { Div } from "lemon-reset";
import apiRoot from "../../apiRoot";

import FeedbackResponseReply from "../FeedbackResponseReply/FeedbackResponseReply";

const ADD_FEEDBACK_RESPONSE_REPLY_MUTATION = `mutation AddFeedbackResponseReply($feedbackResponseId: Int!, $text: String!, $allowReplies: Boolean!) {
    addFeedbackResponseReply(feedbackResponseId: $feedbackResponseId, text: $text, allowReplies: $allowReplies) {
        reply {
            allowReplies
        }
        error
    }
}`;

const FeedbackResponseRepliesTextField = ({
    feedbackResponseId,
    refetchReplies,
    isLoadingReplies,
    allowFurtherReplies,
}) => {
    const [replyText, setReplyText] = useState("");
    const [allowReplies, setAllowReplies] = useState(true);

    const addReply = () => (
        axios.post(`${apiRoot}/graphql/`, {
            query: ADD_FEEDBACK_RESPONSE_REPLY_MUTATION,
            variables: {
                feedbackResponseId,
                text: replyText,
                allowReplies,
            },
        }).then((response) => response.data.data.addFeedbackResponseReply)
    );

    const [addReplyMutate, { isLoading, data }] = useMutation(addReply);

    const onAddReplySubmit = () => {
        addReplyMutate();
    };

    useEffect(() => {
        if (data && data.reply) {
            // Clear textfield
            setReplyText("");
            refetchReplies();
        }
    }, [data, refetchReplies]);

    let canReply = allowFurtherReplies
    if (data && data.reply) {
        canReply = data.reply.allowReplies;
    }

    const onReplyTextChange = (event) => {
        setReplyText(event.target.value);
    };

    const onAllowRepliesChange = (event) => {
        setAllowReplies(event.target.checked);
    };

    return (
        <Div style={{ padding: "0.5em 0" }}>
            {canReply && (
                <Div>
                    <Row gutter={[16, 16]}>
                        <Col>
                            {data && data.error && <Alert message={data.error} type="error" showIcon />}
                        </Col>
                    </Row>
                    <Row gutter={[16, 16]}>
                        <Col>
                            <Input.TextArea
                                placeholder="Write your reply..."
                                value={replyText}
                                onChange={onReplyTextChange}
                                rows={2}
                                autoFocus
                            />
                        </Col>
                    </Row>
                    <Row gutter={[16, 16]}>
                        <Col style={{ display: "flex" }}>
                            <Checkbox
                                checked={allowReplies}
                                onChange={onAllowRepliesChange}
                                style={{ marginRight: "auto" }}
                            >
                                Allow additional replies.
                            </Checkbox>
                            <Button
                                type="primary"
                                loading={isLoading}
                                disabled={!canReply || !replyText}
                                onClick={onAddReplySubmit}
                                style={{ marginLeft: "auto" }}
                            >
                                Send Reply
                            </Button>
                        </Col>
                    </Row>
                </Div>
            )}
            {(!isLoadingReplies && !canReply) && (
                <Typography.Text strong>
                    Additional replies have been disabled for this conversation.
                </Typography.Text>
            )}
        </Div>
    );
}

export default FeedbackResponseRepliesTextField;
