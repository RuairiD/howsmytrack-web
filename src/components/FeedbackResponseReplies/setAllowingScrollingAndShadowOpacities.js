/*
 * Helper function for FeedbackResponseReplies to set container shadow opacities to indicate
 * content outside of scrolled window.
 */
const setAllowingScrollingAndShadowOpacities = (
    repliesDiv,
    setAllowingScrolling,
    setTopShadowOpacity,
    setBottomShadowOpacity,
) => {
    // If there is content that isn't visible because it is being rendered above or
    // below the visible portion of the scrollable container, display the shadows to
    // make it clear there's content that's being hidden. Otherwise, hide them.
    const scrollableHeight = repliesDiv.scrollHeight - repliesDiv.clientHeight;
    if (scrollableHeight > 0) {
        const opacity = repliesDiv.scrollTop / scrollableHeight;
        setAllowingScrolling(true);
        setTopShadowOpacity(opacity);
        setBottomShadowOpacity(1 - opacity);
    } else {
        setAllowingScrolling(false);
        setTopShadowOpacity(0);
        setBottomShadowOpacity(0);
    }
};

export default setAllowingScrollingAndShadowOpacities;
