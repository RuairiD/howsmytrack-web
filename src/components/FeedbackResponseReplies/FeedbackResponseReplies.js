import React, { useCallback, useEffect, useState, useRef } from "react";
import { useQuery, useMutation } from "react-query";
import axios from "axios";

import { Spin, Typography } from "antd";
import { Div } from "lemon-reset";
import apiRoot from "../../apiRoot";

import FeedbackResponseReply from "../FeedbackResponseReply/FeedbackResponseReply";
import AddReplyForm from "./AddReplyForm";
import setAllowingScrollingAndShadowOpacities from "./setAllowingScrollingAndShadowOpacities";

type Props = {
    feedbackResponseId: number,
    feedback: string,
};

const REPLIES_QUERY = `query Replies($feedbackResponseId: Int!) {
    replies(feedbackResponseId: $feedbackResponseId) {
        allowFurtherReplies
        replies {
            id
            username
            text
            timeCreated
        }
    }
}`;

const MARK_REPLIES_AS_READ_MUTATION = `mutation MarkRepliesAsRead($replyIds: [Int!]!) {
    markRepliesAsRead(replyIds: $replyIds) {
        success
        error
    }
}`;

const FeedbackResponseReplies = ({
    feedbackResponseId,
    feedback,
}: Props) => {
    // We don't need to show the scroll bar if there's not enough
    // content to warrant scrolling.
    const [allowingScrolling, setAllowingScrolling] = useState(true);
    const [topShadowOpacity, setTopShadowOpacity] = useState(0);
    const [bottomShadowOpacity, setBottomShadowOpacity] = useState(0);

    const repliesContainer = useRef(null);

    const onRepliesContainerScroll = () => {
        setAllowingScrollingAndShadowOpacities(
            repliesContainer.current,
            setAllowingScrolling,
            setTopShadowOpacity,
            setBottomShadowOpacity,
        );
    };

    const scrollToBottomReply = useCallback(() => {
        repliesContainer.current.scrollTop = repliesContainer.current.scrollHeight;
        onRepliesContainerScroll();
    }, [repliesContainer]);

    const markRepliesAsRead = ({ replies }) => {
        // Mark all replies as read when the modal is first visible.
        // The backend will ensure that only replies sent *to* the
        // logged-in user will be marked as read as obviously marking
        // all replies as read (including the other user's) would be
        // very confusing.
        const replyIds = [];
        for (const reply of replies) {
            replyIds.push(reply.id);
        }
        return axios.post(`${apiRoot}/graphql/`, {
            query: MARK_REPLIES_AS_READ_MUTATION,
            variables: {
                replyIds,
            },
        });
    };

    const [markRepliesAsReadMutate] = useMutation(markRepliesAsRead);

    const getReplies = () => (
        axios.post(`${apiRoot}/graphql/`, {
            query: REPLIES_QUERY,
            variables: {
                feedbackResponseId,
            },
        }).then((response) => response.data.data.replies)
    );

    const { isLoading: isLoadingReplies, data: repliesData, refetch: refetchReplies } = useQuery(
        // By refetching when `addReplyData` changes, we can force
        // the replies to update with the newly added reply.
        [REPLIES_QUERY],
        getReplies,
    );

    useEffect(() => {
        if (repliesData && repliesData.replies) {
            markRepliesAsReadMutate({ replies: repliesData.replies });
            scrollToBottomReply();
        }
    }, [repliesData, markRepliesAsReadMutate, scrollToBottomReply]);

    const allowFurtherReplies = (repliesData && repliesData.allowFurtherReplies);

    return (
        <Spin spinning={isLoadingReplies}>
            <Div className="scrollable-container">
                <Div
                    className="shadow shadow-top"
                    style={{
                        opacity: topShadowOpacity,
                    }}
                />
                <Div
                    onScroll={onRepliesContainerScroll}
                    className={(!isLoadingReplies && allowingScrolling) ? "replies-container" : null}
                    tagRef={repliesContainer}
                >
                    <Typography.Paragraph style={{
                        overflowWrap: "break-word",
                        wordWrap: "break-word",
                    }}
                    >
                        <Typography.Text strong>Original Feedback: </Typography.Text>"{feedback}"
                    </Typography.Paragraph>
                    {repliesData && repliesData.replies.map((reply) => (
                        <FeedbackResponseReply
                            key={reply.id}
                            username={reply.username}
                            text={reply.text}
                            timeCreated={reply.timeCreated}
                        />
                    ))}
                </Div>
                <Div
                    className="shadow shadow-bottom"
                    style={{
                        opacity: bottomShadowOpacity,
                    }}
                />
            </Div>
            <AddReplyForm
                feedbackResponseId={feedbackResponseId}
                refetchReplies={refetchReplies}
                isLoadingReplies={isLoadingReplies}
                allowFurtherReplies={allowFurtherReplies}
            />
        </Spin>
    );
};

export default FeedbackResponseReplies;
