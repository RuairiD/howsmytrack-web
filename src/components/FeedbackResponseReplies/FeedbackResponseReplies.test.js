import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import axios from "axios";
import waitForExpect from "wait-for-expect";
import { act } from "react-dom/test-utils";
import FeedbackResponseReplies from "./FeedbackResponseReplies";

import setAllowingScrollingAndShadowOpacities from "./setAllowingScrollingAndShadowOpacities";

jest.mock("axios");

jest.mock("./setAllowingScrollingAndShadowOpacities");

describe("FeedbackGroupPreview", () => {
    beforeEach(() => {
        axios.post.mockRestore();
    });

    it("displays an empty loading state before replies load", () => {
        const wrapper = shallow(
            <FeedbackResponseReplies
                feedbackResponseId={1901}
                feedback="feedback"
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("loads and renders replies, and marks the replies as read", async () => {
        axios.post.mockResolvedValue({
            data: {
                data: {
                    replies: {
                        allowFurtherReplies: true,
                        replies: [
                            {
                                id: 1901,
                                username: "username",
                                text: "text",
                                timeCreated: 123456,
                            },
                        ],
                    },
                },
            },
        });

        let wrapper;
        await act(async () => {
            wrapper = mount(
                <FeedbackResponseReplies
                    feedbackResponseId={1901}
                    feedback="feedback"
                />,
            );
        });

        await waitForExpect(async () => {
            wrapper.update();
            expect(wrapper.find("FeedbackResponseReply").length).toBe(1);
            expect(axios.post).toBeCalledWith(
                expect.any(String),
                expect.objectContaining({
                    query: expect.stringContaining("replies("),
                }),
            );
            expect(axios.post).toBeCalledWith(
                expect.any(String),
                expect.objectContaining({
                    query: expect.stringContaining("markRepliesAsRead("),
                }),
            );
        });
    });

    it("scrolls to the bottom of the replies container after receiving replies, and set shadow opacities", async () => {
        axios.post.mockResolvedValue({
            data: {
                data: {
                    replies: {
                        allowFurtherReplies: true,
                        replies: [
                            {
                                id: 1901,
                                username: "username",
                                text: "text",
                                timeCreated: 123456,
                            },
                        ],
                    },
                },
            },
        });

        let wrapper;
        await act(async () => {
            wrapper = mount(
                <FeedbackResponseReplies
                    feedbackResponseId={1901}
                    feedback="feedback"
                />,
            );
        });

        await waitForExpect(async () => {
            wrapper.update();
            expect(setAllowingScrollingAndShadowOpacities).toHaveBeenCalled();
        });
    });
});
