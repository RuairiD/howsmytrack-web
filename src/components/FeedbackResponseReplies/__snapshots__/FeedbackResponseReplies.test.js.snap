// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackGroupPreview displays an empty loading state before replies load 1`] = `
<Spin
  size="default"
  spinning={true}
  wrapperClassName=""
>
  <Div
    className="scrollable-container"
  >
    <Div
      className="shadow shadow-top"
      style={
        Object {
          "opacity": 0,
        }
      }
    />
    <Div
      className={null}
      onScroll={[Function]}
      tagRef={
        Object {
          "current": null,
        }
      }
    >
      <Paragraph
        style={
          Object {
            "overflowWrap": "break-word",
            "wordWrap": "break-word",
          }
        }
      >
        <Text
          strong={true}
        >
          Original Feedback: 
        </Text>
        "
        feedback
        "
      </Paragraph>
    </Div>
    <Div
      className="shadow shadow-bottom"
      style={
        Object {
          "opacity": 0,
        }
      }
    />
  </Div>
  <AddReplyForm
    feedbackResponseId={1901}
    isLoadingReplies={true}
    refetchReplies={[Function]}
  />
</Spin>
`;
